File: Java/src/LinkedLists/LinkedListReversalRecursive.java
package LinkedLists;

import DS.ListNode;

/*
    // Definition of ListNode:
    public class ListNode {
        public int val;
        public ListNode next;
        public ListNode(int val) { this.val = val; }
        public ListNode(int val, ListNode next) {
            this(val);
            this.next = next;
        }
    }
 */


public class LinkedListReversalRecursive {
    public static ListNode solution(ListNode head) {
        // Base cases.
        if (head == null || head.next == null) return head;
        else {
            // Recursively reverse the sublist starting from the next node.
            ListNode newHead = Solution(head.next);
            // Connect the reversed linked list to the head node to fully
            // reverse the entire linked list.
            head.next.next = head;
            head.next = null;
            return newHead;
        }
    }
}

File: kotlin/Linked Lists/LinkedListReversalRecursive.kt
import ds.ListNode

/*
    Definition of ListNode:
    data class ListNode(var value: Int = 0, var next: ListNode? = null)
*/

fun linkedListReversalRecursive(head: ListNode?): ListNode? {
    // Base cases.
    if (head == null || head.next == null) {
        return head
    }
    // Recursively reverse the sublist starting from the next node.
    val newHead = linkedListReversalRecursive(head.next)
    // Connect the reversed linked list to the head node to fully
    // reverse the entire linked list.
    head.next!!.next = head
    head.next = null
    return newHead
}
